tincr/io/library/genCellLibrary.tcl

```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+    dict set cell_site_map [get_lib_cells VCC] [list]
+    dict set cell_site_map [get_lib_cells GND] [list]
+        
     return $cell_site_map
 }
 
+## Creates the cell library XML for a leaf cell
 #
-proc process_leaf_cell {c s fo} {
-    create_net tmp_net
-    foreach b [get_bels -of $s] {
-        if { [is_my_placement_legal $c $b] == 1 } then {
-            puts $fo "        <bel>"
-            puts $fo "          <id>"
-            puts $fo "            <primitive_type>[tincr::sites::get_type $s]</primitive_type>"
-            puts $fo "            <name>[suffix $b "/"]</name>"
-            puts $fo "          </id>"
+# @param cell Instance of a library cell to process
+# @param site Site location to place the cell on
+# @param xml_out Output file to print the XML to
+#
+proc process_leaf_cell {cell site xml_out} {
+  
+    # create_net tmp_net
+    set cell_group [get_property PRIMITIVE_GROUP $cell]
+    set cell_can_permute_pins [expr {$cell_group == "LUT" || $cell_group == "INV" || $cell_group == "BUF"}] 
+    
+    # Find all of the configurations of the cell : TODO: add this at a later date
+    if {$::tincr::debug} {
+        set config_list [list]
+        set config_value_map [dict create]
+        set value_count [get_configurations $cell config_list config_value_map]
+    
+        set has_configs [expr {[llength $config_list] > 0}]
+        global config_threshold
+        set above_threshold [expr {$value_count > $config_threshold}]
+    
+        puts "Config count: [llength $config_list] Value count: $value_count"
+    }
+    
+    foreach bel [get_bels -of $site] {
+        if { [is_my_placement_legal $cell $bel] == 1 } then {
+            puts $xml_out "        <bel>"
+            puts $xml_out "          <id>"
+            puts $xml_out "            <primitive_type>[tincr::sites::get_type $site]</primitive_type>"
+            puts $xml_out "            <name>[suffix $bel "/"]</name>"
+            puts $xml_out "          </id>"
 
             #if the placement is legal, place the cell onto the bel to get pin mapping info
-            place_cell $c $b
-
-            #cell to bel pin name dictionary
-            set pin_mappings [dict create]
-            set unconnectedCellPins [list]
+            # place_cell $cell $bel
+            # set unconnectedCellPins [list]
 
-            #Populate the cell pin to bel pin mappings dictionary
-            foreach cell_pin [get_pins -of $c] {
-                set ref_name_cp [lindex [split $cell_pin "/"] end]
+            # create a dictionary of cell pin to BEL pin mappings
+            if {$cell_can_permute_pins} {
+                # set pin_map [create_leaf_cell_pin_mapping_permutable $cell $bel]
+                create_leaf_cell_pin_mapping_permutable $cell $bel $xml_out
+            } else {
+                get_static_leaf_cell_pin_mappings $cell $bel $xml_out
+            }
+            
+            if {0} { ; # START COMMENT
```

Comment:
This line is for configurable cell pin mappings. Currently RapidSmith does not support this, which is why this part of the code is commented out for now. We will need it soon, so I don't want to delete the code. I have added a comment describing why this if statement is there.